### DISCUSSION TITLE
Add redirects for www or non-www

### DISCUSSION THREAD

**Discussion by matthewstick** [craftcms/cloud#12] [Feature Requests & Enhancements]
Usually we just run a site on www or non-www.  We don't want to run the site on both.  I don't see a way for us to "strip wwws" or "add wwws" or "leave both / ignore" in our domains.

Would be great if this was a feature, since it is something we need to do for every site.

In the meantime, I added php code in the index.php to handle this - but this is not ideal.
$host = isset($_SERVER['HTTP_X_FORWARDED_HOST']) ? $_SERVER['HTTP_X_FORWARDED_HOST'] : $_SERVER['HTTP_HOST'];
if (substr($host, 0, 4) === 'www.') {
    $nonWwwHost = substr($host, 4);
    $requestUri = $_SERVER['REQUEST_URI'];
    $newUrl = 'https://' . $nonWwwHost . $requestUri;
    header('Location: ' . $newUrl);
    exit;
}

**Comment by jasonmccallister** [comment]
Hi @matthewstick, this is on our list to have a general redirect offering that would cover this scenario as well. 

We can keep this open for discussion/implementation when the time comes :)

**Comment by svale** [comment]
I'd Just like to second this :) 

Aside from a general redirect functionality, a dedicated option to select a "primary" or "main" domain when adding custom domains would be very nice. 

Ref the Vercel-style setup:

<img width="458" alt="image" src="https://github.com/user-attachments/assets/2c191d59-8b3f-4460-83ff-d48380293892">

**Comment by michtio** [comment]
I want to upvote this as well.

Having both www and non-www available actually gives a negative impact on SEO, marketing agencies don't like it either, google doesn't like it, but currently there is no easy way to redirect www to non-www on a 301.

Bad SEO or duplicate content, can have a very negative impact on the business.

**Reply by timkelty** [reply]
As a best practice, you should always be using [`rel="canonical"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel#canonical), which solves the SEO issue, at least.

**Comment by timkelty** [comment]
This is on our short-list, but in the meantime, you can achieve this via `config/app.web.php` like so:

```php
<?php
// config/app.web.php
return [
    'on beforeRequest' => function (\yii\base\Event $event) {
        $fromHost = 'mysite.com';
        $toHost = 'www.mysite.com';

        if (Craft::$app->getRequest()->hostName !== $fromHost) {
            return;
        }

        Craft::$app->getRequest()->setHostInfo("https://$toHost");
        Craft::$app->getResponse()
            ->redirect(Craft::$app->getRequest()->getAbsoluteUrl(), 301)
            ->send();

        $event->handled = true;
        Craft::$app->end();
    },
];
```

**Reply by jamesmacwhite** [reply]
As a small addition, wouldn't the status code be better as a 301, the default is 302, which is temporary.

e.g.

```php
Craft::$app->getResponse()
            ->redirect(Craft::$app->getRequest()->getAbsoluteUrl(), 301)
            ->send();
```

**Reply by timkelty** [reply]
Thanks @jamesmacwhite – updated my example.

**Comment by jamesbrooks** [comment]
edit: for the before I was applying the change to `app.php` rather than `app.web.php`, make sure you’re applying the change to the correct file! Thanks @angrybrad!

I noticed with the above with craft cloud I would get an issue when building the container (or running any console commands) and had to modify the above block to first include:

```php
$request = Craft::$app->getRequest();

if (!$request instanceof craft\web\Request) {
    return;
}
 ```

(and then optionally just replacing any other calls to getRequest with the already fetched request object).

**Reply by jamesmacwhite** [reply]
Interesting. We recently went live in production and haven't experienced any issues with the code above to handle apex requests conflicting with console or builds. We don't tend to run console commands but use HTTP requests to trigger queue job by schedule due to the lack of native cron scheduling currently

**Reply by angrybrad** [reply]
@jamesbrooks guessing it's because you added it to `config/app.php` (applies to both web and console requests) and not `config/app.web.php` (which only applies to web requests).

**Reply by jamesbrooks** [reply]
@angrybrad yes that’s exactly it! I missed that difference from the original comment, sorry! I’ll edit my comment. Thanks!

**Comment by timkelty** [comment]
This feature is currently being beta tested.
If you'd like your project opted-in, please contact support@craft.cloud with the project/environment you'd like to try it on.
